Clean up various issues surrounding `ScoreInfo` in preparation for realm migration

peppy opened the pull request:
These are pieces I am able to pull out of my ongoing work. Recommend reviewing on a per-commit level – changes are grouped.

peppy committed:
Change `IScoreInfo.User` to an interface type

peppy committed:
Use `OnlineID` instead of legacy IDs for equality and lookups

peppy committed:
Avoid touching `ScoreInfo.User` directly

peppy committed:
Add setter for `ScoreInfo.OnlineID`

peppy committed:
Use `OnlineID` for set operations

peppy committed:
Rename `APIScoreInfo` to `APIScore`

frenzibyte commented on osu.Game/Scoring/ScoreInfo.cs:
```diff
@@ -237,8 +238,8 @@ public bool Equals(ScoreInfo other)
 
-            if (OnlineScoreID.HasValue && other.OnlineScoreID.HasValue)
-                return OnlineScoreID == other.OnlineScoreID;
+            if (OnlineID > 0)
```
The online ID of the other score should matter as well?
```suggestion
            if (OnlineID > 0 && other.OnlineID > 0)
```

peppy replied:
I'm actually not even sure if we want this here. It doesn't match the new equality methods we decided on for `Beatmap` and `BeatmapSet`. Let me see if I can isolate usages of this and remove it instead.

peppy committed:
Remove `OnlineID` comparison from `ScoreInfo.Equals`

This matches the implementation we have for `BeatmapInfo` and
`BeatmapSetInfo`. All comparisons of `OnlineID` should be done directly
using them (ie. how it's done in `ScoreModelDownloader`).

peppy committed:
Add `MatchesOnlineID` implementation for `IScoreInfo`

peppy committed:
Use similar method of consuming `OnlineID` as done in beatmap classes

smoogipoo committed:
Merge branch 'master' into pre-realm-score-tidy

smoogipoo commented:
Fyi I have ongoing changes to fix test failures, unrelated to this/other branches of course.
